sympy/solvers/tests/test_solveset.py

```diff
@@ -2114,4 +2149,12 @@ def test_solve_modular():
             Intersection(S.Complexes, ImageSet(Lambda(n, \
             3*sqrt(11)*sqrt(21691754*n + 1)), S.Integers)))
 
+@XFAIL
+def test_solve_modular_fail():
+    # this test is passing but for wrong reason because both are
+    # getting evaluated to EmptySet() which is wrong because Mod for complex
+    # is not defined yet.
+    assert solveset(Mod(x, 3) - I, x) == \
+            ConditionSet(x, Eq(Mod(x, 3) - I, 0), S.Complexes)
```

Comment:
XFAIL means the test is expected to fail, so this will show up at each run as a test that unexpectedly passed. (But it's not an XPASS either since it's not intended to fail.)

As for the actual test, I tracked down the ConditionSet behaviour:
```
In [26]: Mod(x,3).is_real
Out[26]: True
```
So `Eq` immediately evaluates to false since a real can't equal i. For this particular equation I'd say that the `EmptySet` output of `_solve_modular` and hence that of `solveset` is technically correct. 

Though I'm wondering if `domain=S.Complexes` should not be excluded from the get-go. Let's see what other reviewers think.

(Maybe you could just leave the test as `assert ... == S.EmptySet`
and add a note that `Mod` is inherently real, as are its two arguments.)
```

Still, there are more domain issues remaining:
```
In [32]: _solve_modular(Mod(I+x, 3), x, S.Reals)
Out[32]: {3⋅n + 2⋅ⅈ | n ∊ ℤ}
```

Reply:
I dont get that why this is happening.
```
>>> Mod(I+x, 3).is_real
True
```
I think it should return False.

Reply:
`_solve_modular(Mod(I+x, 3), x, S.Reals)` is now returning a ConditionSet. Good!

As for `Mod(I+x, 3).is_real` returning `True`, consider this: `Mod` is currently only defined for reals (except for zero which raises an exception) in the sense that in all other cases it remains unevaluated. So *if* it evaluates it's *guaranteed* that the result will be a real number. That seems to be the current logic.

`False` would technically be wrong: just consider `x = -I`, `x = 7-I`, etc. In any case, Mod and Complexes don't mix well.

Reply:
Then it should return `None` if thats the case.